modules/core/src/main/java/org/apache/synapse/mediators/template/TemplateContext.java

```diff
@@ -92,7 +93,11 @@ private Object getEvaluatedParamValue(MessageContext synCtx, String parameter, V
                 	}
                     return expression.getExpression();
                 } else {
-                    return resolveExpressionValue(synCtx, expression);
+                    if (expression.getExpression() instanceof SynapseJsonPath) {
```

Comment:
I think overriding resolveObject in SynapseJsonPath will be a cleaner fix.

Reply:
If we override selectSingleNode() in SynapseJsonPath, then it has to return a String value in selectSingleNode. But method name describe itself as a SingleNode. So we need to refactor the method names as well. 
+1 for refactoring whole path of evaluating the parameters in template. Then it will be more clear.